libs/sync/rw_init.go

```diff
@@ -0,0 +1,95 @@
+package sync
+
+import (
+	"sync"
+)
+
+// RWInitMutex is a RWMutex that also has the notion of an initialized state
+// that writers can set, and readers can wait for.
+type RWInitMutex struct {
+	*RWMutex
+	writeLocked bool
+	initialized bool
+	waitForInit *sync.Cond
+}
+
+func NewRWInitMutex() *RWInitMutex {
+	mtx := new(RWMutex)
+	rwi := &RWInitMutex{
+		RWMutex: mtx,
+		// Condition variables should work with read locks, as long as the actual
+		// state update is done under the associated write lock to provide
+		// happens-before visibility. See comment for notifyListAdd() in
+		// src/runtime/sema.go in the Go sources
+		waitForInit: sync.NewCond(mtx.RLocker()),
+	}
+
+	rwi.waitForInit = sync.NewCond(rwi.RLocker())
+	return rwi
+}
+
+type readInitLocker struct {
+	sync.Locker
+	rwil *RWInitMutex
+}
+
+// RInitLocker creates a locker that waits until the rwi.Initialize() has been
+// called, then holds a read lock.
+func (rwi *RWInitMutex) RInitLocker() sync.Locker {
+	return readInitLocker{
+		Locker: rwi.waitForInit.L,
+		rwil:   rwi,
+	}
+}
+
+func (rwi *RWInitMutex) RInitLock() {
+	rwi.RInitLocker().Lock()
+}
+
+func (rwi *RWInitMutex) RInitUnlock() {
+	rwi.RInitLocker().Unlock()
+}
+
+func (ril readInitLocker) Lock() {
```

Comment:
For the RInitLocker stuff (lines 31-68), consider the following which might be more direct:

```
func (rwi *RWInitMutex) RInitLock() {
    rwi.mtx.RLock()
    while !rwi.initialized { // easier to just loop rather than explain why we're not
        rwi.waitForInit.Wait()
    }
}

func (rwi *RWInitMutex) RInitUnlock() { rwi.mtx.RUnlock() }

type rInitLocker RWInitMutex
func (r *rInitLocker) Lock() { (*RWInitMutex)(r).RInitLock() }
func (r *rInitLocker) Unlock() { (*RWInitMutex)(r).RInitUnlock() }

func (rwi *RWInitMutex) RInitLocker() sync.Locker {
    return (*rInitLocker)(rwi)
}
```

Reply:
Love it!  Thanks!